File: internal/models/monitoring.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// MonitoringRecord represents a system monitoring record saved to database
type MonitoringRecord struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at" gorm:"index"`

	// CPU metrics
	CPUUsagePercent float64 `json:"cpu_usage_percent"`
	CPUCores        int     `json:"cpu_cores"`

	// Memory metrics
	MemoryTotal       uint64  `json:"memory_total"`
	MemoryUsed        uint64  `json:"memory_used"`
	MemoryAvailable   uint64  `json:"memory_available"`
	MemoryUsedPercent float64 `json:"memory_used_percent"`

	// Disk metrics
	DiskTotal       uint64  `json:"disk_total"`
	DiskUsed        uint64  `json:"disk_used"`
	DiskFree        uint64  `json:"disk_free"`
	DiskUsedPercent float64 `json:"disk_used_percent"`

	// Network metrics
	NetworkBytesSent   uint64  `json:"network_bytes_sent"`
	NetworkBytesRecv   uint64  `json:"network_bytes_recv"`
	NetworkPacketsSent uint64  `json:"network_packets_sent"`
	NetworkPacketsRecv uint64  `json:"network_packets_recv"`
	NetworkSpeedSent   float64 `json:"network_speed_sent"` // bytes per second
	NetworkSpeedRecv   float64 `json:"network_speed_recv"` // bytes per second

	// Host info
	Hostname string `json:"hostname"`
	Uptime   uint64 `json:"uptime"`
}

// BeforeCreate hook to set created_at if not set
func (m *MonitoringRecord) BeforeCreate(tx *gorm.DB) error {
	if m.CreatedAt.IsZero() {
		m.CreatedAt = time.Now()
	}
	return nil
}

File: frontend/src/types/monitoring.ts
// 系统监控相关类型定义

export interface CPUStats {
  usage_percent: number;
  cores: number;
  per_core: number[];
}

export interface MemoryStats {
  total: number;
  used: number;
  available: number;
  used_percent: number;
}

export interface DiskStats {
  total: number;
  used: number;
  free: number;
  used_percent: number;
}

export interface NetworkStats {
  bytes_sent: number;
  bytes_recv: number;
  packets_sent: number;
  packets_recv: number;
  /** bytes per second */
  speed_sent: number;
  /** bytes per second */
  speed_recv: number;
}

export interface HostStats {
  hostname: string;
  os: string;
  platform: string;
  platform_version: string;
  uptime: number;
  boot_time: number;
}

export interface SystemStats {
  cpu: CPUStats;
  memory: MemoryStats;
  disk: DiskStats;
  network: NetworkStats;
  host: HostStats;
}

export interface MonitoringRecord {
  id: number;
  created_at: string;
  cpu_usage_percent: number;
  cpu_cores: number;
  memory_total: number;
  memory_used: number;
  memory_available: number;
  memory_used_percent: number;
  disk_total: number;
  disk_used: number;
  disk_free: number;
  disk_used_percent: number;
  network_bytes_sent: number;
  network_bytes_recv: number;
  network_packets_sent: number;
  network_packets_recv: number;
  network_speed_sent: number;
  network_speed_recv: number;
  hostname: string;
  uptime: number;
}

export interface MonitoringHistoryResponse {
  records: MonitoringRecord[];
  count: number;
}

export interface ChartDataPoint {
  timestamp: number;
  cpu_percent: number;
  memory_percent: number;
  disk_percent: number;
  network_speed_sent: number;
  network_speed_recv: number;
}

export interface MonitoringChartResponse {
  points: ChartDataPoint[];
  count: number;
  period: {
    hours: number;
    from: number;
    to: number;
  };
}

export interface MonitoringStatsResponse {
  records: MonitoringRecord[];
  stats: {
    count: number;
    period?: {
      hours: number;
      from: number;
      to: number;
    };
    cpu?: { average: number; max: number };
    memory?: { average: number; max: number };
    disk?: { average: number; max: number };
  };
}
